### ISSUE TITLE
More issues

### ISSUE THREAD

**Issue by smblackledge** [thonny/thonny#1228]
Features I really do use in other debuggers (e.g. Visual Studio).  It would be nice if the debugger supported:

- watchpoints

- triggering a breakpoint/watchpoint after the nth time.  If code fails the 307th time through a loop, skip the first 306.

- displaying numbers in other bases.  For example, hex and binary

- displaying line numbers in the debug window

- setting another breakpoint when at an existing breakpoint

- ERROR - clicking above or below the movable thingy in a slider widget doesn't do anything

- I'm reporting this without trying it again, but I remember using "sys.exit(main())" at the end of a program.  This causes a traceback to be displayed (something wasn't handled correctly) when single-stepping but it executes correctly at full speed.

- ctrl-? would cause a keyboard interrupt which would enter something like a breakpoint.  This would allow variables to be displayed/examined, etc.  You would have to have started running with the debugger.  I'm not suggesting hitting the green arrow to start running and then expecting to trigger a breakpoint.  I'm guessing that you'd have to be running with the "nice" or "fast" debugger selected.

- I'm running a practice problem from Topcoder (ABBAReplace problem).  This generates strings that are 7M in length.  Thonny will hang.  I don't think it is running out of memory.  I don't have any idea why the hang, though (ran out of some resource?).  Here's the source (what I was debugging when I encountered most of the items in this email).  Please name it ABBAReplace.py.  It will run a *long* time.

Scott

```
################################################################################
# Consider a very simple string rewriting system. We have a string over the
# alphabet {A,B}. In each step, we find all occurrences of "AB" in the current
# string and we change each of them into "BA".
# 
# For example, if we start with the string "AABABB", the rewriting will go
# as follows:
# 
# 0: AABABB
# 1: ABABAB
# 2: BABABA
# 3: BBABAA
# 4: BBBAAA
#
# At this point, the rewriting has stopped, as there are no more "AB"
# substrings in the current string.
# 
# Use the following pseudocode to generate a string S of length N:
# 
# state = seed
# S = Sprefix
# while length(S) < N:
#     state = (state * 1103515245 + 12345) modulo 2^31
#     if state < threshold:
#         S += 'A'
#     else:
#         S += 'B'
#
# Given the string S, compute and return the number of rewriting steps.


testcases = [
#        Sprefix, N, seed, threshold, expected
        ["AABABB", 6, 0, 0, 4],
        ["", 0, 4, 7, 0],
        ["ABBABAABABBBABBBB", 17, 0, 0, 11],
        ["AABAA", 17, 47474747, 1000000000, 10],
        ["", 7000000, 47, 1123456789, 3661759],
        ["", 7000000, 47, 1987654321, 6479030],
        ["", 7000000, 47, 471, 5849431],
        ["B", 1, 1847635854, 2147233648, 0],
        ["B", 2, 1797583853, 2196388, 0],
        ["ABB", 3, 1355395163, 486381, 2],
        ["ABBA", 4, 1607036031, 2140324368, 2],
        ["AAAAB", 5, 1558380042, 2147020398, 4],
        ["AABAB", 6, 1516333974, 3298058, 4],
        ["BBABBBB", 7, 489349967, 2146000780, 4],
        ["BBABA", 8, 2046595582, 2084695727, 1],
        ["ABBBAAABB", 9, 758823778, 45788, 5],
        ["AABBABBAAB", 10, 1585910533, 1, 6],
        ["AABABBAAAAB", 11, 1007175254, 2142775666, 7],
        ["AAABABAABABB", 12, 492182829, 2147468613, 8],
        ["BBAAABBABABBA", 13, 1512640368, 55774824, 7],
        ["BAABBBAABABBBA", 14, 2050067571, 2147257689, 8],
        ["BAAABBBBAAB", 15, 1150680127, 2147483284, 7],
        ["ABBBBBAABAABBBBB", 16, 1301697396, 2147483637, 11],
        ["BBBABBABBABAAAABB", 17, 963875836, 2147461499, 8],
        ["AAABABBBBABAABB", 18, 402545275, 2147476869, 10],
        ["ABBAABBABBABABAAABB", 19, 268214787, 2147482954, 10],
        ["ABBABABABAAAAAAAAAAA", 20, 1691293679, 227717151, 5],
        ["BBABABAAAABABBBBBBBBA", 21, 1402024454, 3, 14],
        ["BA", 22, 27067550, 432, 20],
        ["ABBAABA", 23, 1792312856, 2147481236, 3],
        ["BBAABBAAAABBABAAABABBABB", 24, 1781210239, 1729, 14],
        ["AABAAAAABBAABB", 25, 1604168383, 2147483633, 10],
        ["ABAAAAAAABBBBAABBBAAAAAAAA", 26, 546676137, 2146653054, 14],
        ["BAAAA", 27, 2007325964, 4926, 25],
        ["ABBAABBAABBBABBBBAAAABABBABA", 28, 763289261, 2147102227, 15],
        ["AAABBABAABABBABAAAB", 29, 164618037, 2123745304, 11],
        ["BBAAABBBABAABAABBABAAABBABAAAA", 30, 1800803545, 2147482813, 14],
        ["ABBAAABBAABBBBA", 6912249, 326077955, 490, 6912242],
        ["BBBBABAAAAAABBABBABBAABAABBBBABBBBABABBBBBAAABABAABABABBBABBAAABB", 6075671, 803987253, 2147483508, 5622913],
        ["AABBAAABBABBBBBAABAAAABABABABABBAAAAAAABABAABBABABAAAABBB", 6410758, 112917720, 705, 6410732],
        ["BBABABABAAAAABBBBABAAB", 5903397, 1314441501, 2147483444, 13],
        ["BBBBABAABBABBABBABBBBA", 5375659, 1623874294, 10494, 5375616],
        ["AAABABBBABAAAABBABBBBAABBAABABBABAABABBABBABAAAABABBAAA", 6540593, 890467798, 2147483636, 29],
        ["BABBBAABBAAABAAAABABAABBABABAABBABABAB", 5803744, 830502508, 5632, 5803718],
        ["ABBBBBAAAABBBBBABBABABBBAAAABBABBAAAAABAAABBAAABBBABBABBBBABAAA", 5502432, 421874798, 2147483644, 33],
        ["ABAAABBBABABABBAAAABAABABBBBAABBAAAAABAAAABBBBAABBABBBAABBBABABAABAAAABABABAAABABBAB", 5432079, 1590457026, 336661834, 4580826],
        ["BABAABAAAAAAAAABAABAAABAABAAABABAAABABABBAABAAABBBBA", 5898103, 1578742758, 12610551, 5863389],
        ["AAABABBBAABBAABABBAABBBABBBABABBAABAAAABABBAAABBABBBBAABABBBBABABBBAAAAABBBAAABAABBABAAAAAABBBABAABB", 6091081, 1759666786, 2147025113, 6087399],
        ["ABBBAABABABBAAABAAABAABBABABAAABABAA", 5512478, 1688072213, 2997, 5512450],
        ["BBBBBBAAAABBAAAAAAAABAAAA", 5523978, 1475271704, 2134851497, 5491228],
        ["ABBAABAAABABBAABABABBBABABAABBAABAABABABAABBABABAAAABABAAABBBBBBBAABBBBBAABBABBABAAABBABABAABAA", 5869046, 1695650247, 2142948831, 5856122],
        ["BBBAAAAABBBAABBBBBABAABBABBBBABBBBBAABBBABBABBAAAAABBBAABABAABAABAAAB", 6164614, 151396815, 505071775, 4715511],
        ["ABAAAAAAAAABBBAAAABABBBBABABABBAABAAAAABAAAABBAAABBBBBBABBBBABBBBAAAABBBBBAAABABBBBBAAAAAABB", 6616493, 1304793850, 2147483590, 58],
        ["AABAAABABAAAAAABBAAABBAABABBBAAABBBAAAABBBABBAAAA", 6232402, 2046038195, 2093151009, 6073847],
        ["BBABABAABBAABAAAABBBBABBAAAAAAAAAAABBAABBAAABABAABBAABBAABBBBABB", 5970309, 2075052142, 2147091451, 5958293],
        ["BBBBBBAAABAABABBABABABAAAABABBABBABABBBABBABAABBABBBBAAABB", 6925367, 386610351, 2147299922, 6922293],
        ["AABBBBBBBAAABBBBBBBABAABBBABBBABABAABAABBBBBBABBAABABBBABAABABBBBABAB", 6067653, 2064249094, 2147482116, 5898796],
        ["BBABAAABAABBBABABBBBABBBBABBBAABAABABBAAAABABBAABBBAABBAABAABBBABBABBB", 5585945, 732323548, 2147483646, 41],
        ["BBAAABAABBBABBBBAAAABAA", 5151348, 1796192107, 2125004421, 5097651],
        ["BBBBABABAABAAAAABABBBBBABBBABABBBAABAAABABBABBBABBAAABABAB", 5607503, 1248574835, 2147397927, 5604420],
        ["BBAAABAABAABAAABBBBAAAAAAABAAAABBBBBBAABBBABABAABAAAAAABABBBAABBAABBAAABBABBBBABABBABABB", 6960825, 1931908749, 54398, 6960624],
        ["AAABBBAAABBBAABBABAAABAABAAAAABBABBABBBAAABB", 6424631, 302019225, 3, 6424614],
        ["ABABBB", 5258641, 904693122, 27342, 5258582],
        ["AAAAABBABBBABBBBABBABABAAAAABBAAABAABBBBBBBB", 5357461, 1089999389, 2147128224, 5356206],
        ["ABABABAAABAAAAABABABBABBBAAAAABBBBBABABAAABBBBBBBBABAABBBBBBBABABABAA", 5486452, 1654387811, 2147228962, 5472600],
        ["BBBBABBBABABABABBBAABAABBBABABAAABBBABABBABABBABBBBBABBBABAABABBBAAAB", 6054375, 589686326, 472525, 6053022],
        ["BABAABABBABBABBAAAAAAAABAAABAABABBAABAABBBAAABAABAABAB", 5803462, 1900850002, 2147483394, 32],
        ["BBAABBAAAABBAAAABBABABABAABABBBBABAAABBAAABBABBABAABBAAABAABBAABBBBABB", 6580238, 741745324, 2146865422, 6575156],
        ["BABBBBBBABABAABBBBBAABA", 5565493, 1248462140, 61776714, 5405393],
        ["BAAAAAABABAAABA", 6016538, 2057909502, 48, 6016533],
        ["AAAABABAAAABAAABBAABBAA", 5082833, 1952484420, 2147480744, 4906228],
        ["AABABABBAABBAABAABBABABAAB", 6145504, 1816392216, 348304, 6144461],
        ["BABAA", 5403400, 217274729, 2147483632, 1],
        ["BAABABABBABBAAB", 6855444, 915701050, 15207, 6855396],
        ["AABABBABABAABBABAAAAABBAABBBABBABABAABAABABBBABBABABAABABBBABBAABBAAAABABBBB", 5799899, 1179698765, 2147483257, 2244014],
        ["BABBABBAABAB", 6924594, 1662781606, 382020800, 5691694],
        ["BAAABAAAAAAAAABBBBABBBBABBBABBBBBBBABBBABBABBAABBAAABBBAAAAAABAAAABAABABBAAABAABBBAA", 6264039, 598364207, 245803, 6263237],
        ["BAAAABAABABBAABABBAABAABABBBAABBABAAABBBBBABBAABBABBBAABBBBBABABBBBAAAABBAABBBB", 5573249, 795952589, 2147483643, 48],
        ["BAABAABBBAAAAAAABAABAABBBBBBBBBAAABAABAAAAABBBBABABBBABBBAABABABA", 6430155, 298062256, 2147483646, 39],
        ["BAAABBABAAABAABABBAAABAABBAABABBAAABBAABAABABABBAABBAAABBBBABABBBBAAAAAABAABBABAABABBBBB", 6198758, 616629097, 128665718, 5827622],
        ["AABBABBBBBAAAAABABABABBBABBABBBBBBABABBBABAAAABBAABABAABABABBBAAAABAAB", 6827548, 542765220, 226398, 6826749],
        ["AABABABABBBABBBABABABBBAAAB", 5841104, 1770758136, 216059556, 5253252],
        ["BBAABABAAABBAAAAAAABABBAAAAABABBBBBABABABBBBBABABABBBBAAABABAABBABABBAAABBBAABA", 5378029, 81932240, 105906, 5377735],
        ["ABB", 6589056, 1570009844, 2146169071, 6584409],
        ["ABABBA", 6696991, 724530287, 198848, 6696346],
        ["BABBABAABAABBBBAAAAAABABBBBABBBAABAAABBABBBBBBBBBAABBBABBABBBBABABAAAAABAAAAAABAA", 5310493, 961791610, 2147387337, 5301277],
        ["ABBABAAAABBBAABBAABBABBAABABBABBABABABBBABABAAAAABABABBBBBAABBABAAAABBABABBBABAAABAABAAABBAABBAA", 5136044, 2000846115, 2146110787, 5131721],
        ["BBBAB", 5, 0, 0, 1],
        ["ABAAA", 5, 4, 4, 1],
        ["BBBB", 4, 0, 0, 0],
        ["AABAA", 6000000, 47474747, 1000000000, 3205457],
        ["BBAA", 4, 0, 0, 0],
        ["AABAA", 7000000, 47474747, 1000000000, 3739188],
        ["BBBB", 4, 123, 123, 0],
        ["BBBBBAAAAA", 10, 123123123, 321321321, 0],
        ["BBBBBBBBBB", 10, 0, 0, 0],
        ["BBBAAA", 6, 1, 1, 0],
        ["BBAA", 4, 1010, 1010, 0],
        ["BBBBBBAB", 8, 123, 123, 1],
        ["", 7000000, 4, 7, 0],
        ["BBBAB", 5, 1, 1, 1],
        ["B", 1, 0, 0, 0],
        ["BBBBBBBBBBAB", 12, 12, 100000, 1],
        ["BABBBBAABBABABBAAAAABABBAAABBBBAAABBBABABBBBABAABABABAABAAABBBABABABBBABABABAABABAAAAABBABAAAABAAAABBAAABBBABAAABAABBBABABBBBBBBBBBABBABABBAABABBBAAABABABABABBBBAAABABAAAAABABBBBBAABBABBABABBBBBAABBABAABBBAAAAAAABBBAABABABABAABBAABAAABBBAABAABBAAAABABBBAABAAAAABAAAAABAAAAABBAABBBBABAABBBBABBBAABAAAAAAAAABBABABAAABBAAABABBABBBBBABBABBABAAAAAABABAAAAAABBABAAABAAAABBBAABAABBBAABAABABBBBABBAABAAAAABAAABAAAAAABABBBAAABAABAAABBABBBBBAABAABABBAAABAAABAAAAABBBBAABABBABBBABABBABBABBBBBAABBBBAABBABAAABBAABAABBBBBABABBAAAABAABAAAAAAAAAABAABABABBBBAAABAAABABBABBABBAAAAAABAABBBBBABBBABBBBBABABBABAABAABBBBAABBBBBAABBBAAAABBBBBABBBBBBAAAABBAAABBAAAAABAAABBBAAABBBBABBBAAAABBBABBABBBABABABAAABABAABABBAABBBABABAAABABBBBBBABBAABABBAAAABBBBBBABBAABBABBBABABBABBAABAABAAABBAAABAABAAABBAABABABAABAAABAAAABAABBABAAAABBBBBBABABBBBBBAABAAAAABAABABBBAAAABBAABBBAABABBBBBAABBABAAABBBABBBABABBBABAAABBAABABBABBABBAABBBAABABAABBABAABAAAAABABBAAAABBAABABABBABBAABABBBABBBBAABAAABBAABABBBBBAABBBAAABABAAAAABABBAABABBBABAB", 7000000, 2443, 1000000000, 3739428],
        ["BB", 2, 0, 0, 0],
        ["BBB", 3, 0, 0, 0],
        ["AABBBABB", 8, 1, 1, 6],
        ["B", 7000000, 0, 0, 0],
        ["BAA", 1281, 16827, 4893846, 1275],
        ["", 1000, 520343481, 2147483646, 0],
        ["BBBAAA", 6, 0, 1, 0],
        ["BBBBB", 5, 0, 0, 0],
        ["ABAABBABAA", 10, 1, 1, 5],
        ["AAAAABBAAAAAAAB", 15, 0, 0, 12],
        ["B", 10, 0, 0, 0],
        ["BBBB", 4, 1, 1, 0],
        ["BBBBBBBAABAAABABAB", 7000000, 100, 987654321]
    ]


class ABBAReplace:
    def countSteps(self, Sprefix, N, seed, threshold):
        #                prefix  int int   int
        
        state = seed
        slen = len(S)
        while slen < N:
            state = (state * 1103515245 + 12345) % 2**31
            if state < threshold:
                S += 'A'
                slen += 1
            else:
                S += 'B'
                slen += 1
                
            if (slen % 250000) == 0:
                print(slen)
 
        count = 0
        i = 0
        go = True
        
        print(f'Sprefix = "{Sprefix}", N = {N}, seed = {seed}, threshold = {threshold}')
        print(f'S = "{S}"')
        
        if not S:
            go = False
            
        while go:
            while i < len(S) - 1:          # only look where there could be an AB
                loc = S.find("AB", i)      # find next AB
                if loc >= 0:               # if found
                    S = S[:loc] + "BA" + S[loc + 2:]  # change AB to BA
                    i = loc + 1
                i += 1
            else:
                count += 1
                if not count % 250000:
                    print("Count = {count}")
                
                i = 0
                loc = S.find("AB", i)      # Anything left to do?
                if loc < 0:
                    break
            
        return count    


def main():
    cases = 0
    cl = ABBAReplace()
    for Sprefix, n, seed, threshold, expected in testcases:
        
        cases += 1
        print(f"cases = {cases}, expected = {expected}")
        
        result = cl.countSteps(Sprefix, n, seed, threshold)
        if expected != result:
            print("\nError")
            print(f"Expected {expected}, received {result}")
            break
    print("Done")
    return result

    
if _name_ == '_main_':
    main()
```

**Comment by smblackledge** [comment]
I'm not confident that garbage collection is being done (or being done correctly).  That may be related to why Thonny hangs.  The strings that the program creates are huge and it creates a lot of them.  (Yes, the actual solution does not actually change occurances of AB to BA.  I've got another solution that works.

Sorry for the weird formatting in the email.  It is late and I'm tired.  I can put the program in a file and attach it if you want.

Scott

**Comment by smblackledge** [comment]
I was going to try to fix the formatting by editing the post.  That window showed the code correctly.  Oh well, please tell me if there are problems with the code and I'll quickly get a correction to you.

**Comment by aivarannamaa** [comment]
> watchpoints

I haven't used watchpoints myself. It looks like a feature for advanced users, but I still want to understand it better. As far as I understand, this could be done such that the user provides an expression and the debugger tries to evaluate it on **each** possible stopping point. Is the expression always evaluated against the global environment of the main module? Or against whatever environment the execution is in currently? When the execution moves from global scope to local, the meaning of the expression can change and therefore its value can change as well -- should the debugger stop in this point? When the execution moves to another scope and the expression is not valid anymore, does this count as change in the value?

> triggering a breakpoint/watchpoint after the nth time. If code fails the 307th time through a loop, skip the first 306.

Maybe allowing general conditions on breakpoints would be more useful? In either way, it would be advanced feature and I'm not sure Thonny's target group needs this.

> displaying numbers in other bases. For example, hex and binary

Good idea and easy to implement! I created separate issue for this: #1230

> displaying line numbers in the debug window

Do you mean line numbers in the function call windows? This should be done indeed (https://github.com/thonny/thonny/issues/1231)

> setting another breakpoint when at an existing breakpoint

This should work. Have you experienced problems with this?

> ERROR - clicking above or below the movable thingy in a slider widget doesn't do anything

Don't understand what are you talking about.

> I'm reporting this without trying it again, but I remember using "sys.exit(main())" at the end of a program. This causes a traceback to be displayed (something wasn't handled correctly) when single-stepping but it executes correctly at full speed.

Please create separate issue with example code when you can reproduce this.

> ctrl-? would cause a keyboard interrupt which would enter something like a breakpoint. This would allow variables to be displayed/examined, etc. You would have to have started running with the debugger. I'm not suggesting hitting the green arrow to start running and then expecting to trigger a breakpoint. I'm guessing that you'd have to be running with the "nice" or "fast" debugger selected.

I suppose we could call it "Pause". This has been suggested before, but now it occurred to me that it might be useful for trying to understand why the program ended up in an infinite loop. I created a separate issue: https://github.com/thonny/thonny/issues/1232

> I'm running a practice problem from Topcoder (ABBAReplace problem). This generates strings that are 7M in length. Thonny will hang. I don't think it is running out of memory. I don't have any idea why the hang, though (ran out of some resource?). Here's the source (what I was debugging when I encountered most of the items in this email). Please name it ABBAReplace.py. It will run a long time.

Thonny does hinder garbage collection, because it keeps references to certain objects to be able to display the Heap table. I'll see what can be done (https://github.com/thonny/thonny/issues/1233)

**Comment by smblackledge** [comment]
I’m going to clone the source which will help when writing a user guide.

 

I’ve got to start numbering these requests when I initially ask for something :).

 

I’ll start by suggesting a quick way to talk about what watchpoint environments; until we have something better, I’ll use a letter: the global environment (g), the current thread (t), whatever is using the current stack frame (f), or maybe even the current block of code (b).

 

#1) Would the watchpoint be considered global or local?  If a global resource is being watched, then the expression is evaluated at every stop.  If a local resource, only evaluate the expression when the correct environment is in effect.  If you can’t determine locality, treat the resource as global.  Scope change does not mean the watch has been triggered.  Is this answer good enough?  If no, I can be more thorough.

 

#2) Please reconsider (only break on the nth occurrence).  This won’t be used often, but it will be used.  For breakpoints, I don’t consider it an advanced feature.  This is very useful in a loop where you may or may not know which iteration has the problem.  It is probably an advanced feature when used with watchpoints, though.

 

#4) Yes.

 

#5) No, it doesn’t work.  I set a breakpoint and ran the debugger.  The debugger then stopped at the breakpoint.  Double-clicking on another code line in the code window doesn’t do anything, especially it does not set a new breakpoint.  I don’t know if this problem has to do with the line numbers not being displayed and therefore not having a prefix area to click in to (re)set a breakpoint.

 

#6) Sorry.  I was too tired to write that one and I did too poor a job of explaining.  Hopefully this attempt is better.

 

The slider widget on the right side of the code editor window has the up- and down-arrows and the “bar”, which you can left-click and drag to view different places in the source code.  The arrows and the bar both work, but clicking above the bar and below the bar, thinking I can get about a page’s movement, doesn’t do anything. 

 

#7) I wrote a small test program but cannot get it to fail.  I’ll ignore this for now.

 

#8) That’s exactly the reason I wanted the keyboard to trigger a breakpoint.

 

#00) I’m reviewing Step Back in the document tonight.

 

Scott

**Comment by aivarannamaa** [comment]
> I’ve got to start numbering these requests when I initially ask for something :).

It's better if you create separate issue for each topic. I now created some new issues so please write your comments under these.

> If a global resource is being watched, then the expression is evaluated at every stop.  If a local resource, only evaluate the expression when the correct environment is in effect.

I created a new issue for discussing watches: https://github.com/thonny/thonny/issues/1234

> Please reconsider (only break on the nth occurrence)

Here's a new issue regarding breakpoint properties: https://github.com/thonny/thonny/issues/1235 . Please provide an example which demonstrates the utility of the feature you proposed.

> Double-clicking on another code line in the code window doesn’t do anything, especially it does not set a new breakpoint.

Double-clicking on a code line wouldn't set a breakpoint even before starting the debugger, you need to set it on the line number margin. For local context this depends on #1231 . As a workaround you can turn off "Tools => Options => Run & Debug => Show function calls in separate window".

> The slider widget on the right side of the code editor window has the up- and down-arrows and the “bar”, which you can left-click and drag to view different places in the source code.

I see you meant the scrollbar. Its behavior is defined by the GUI toolkit (Tk). I now confirmed that what you tried really doesn't work on Linux but it does work in Windows. Hopefully next Tk version will fix this for Linux as well

**Comment by kilitary** [comment]
is result hat is to ruin the cheaters?
